components/brave_vpn/brave_vpn_service_desktop.cc

```diff
@@ -427,18 +466,23 @@ void BraveVpnServiceDesktop::LoadPurchasedState() {
   }
 #endif
 
-  const std::string credential =
-      prefs_->GetString(brave_rewards::prefs::kSkusVPNCredential);
-  if (skus_credential_ == credential)
-    return;
-
-  skus_credential_ = credential;
+  const bool has_credential =
+      prefs_->GetBoolean(skus::prefs::kSkusVPNHasCredential);
 
-  if (!skus_credential_.empty()) {
-    VLOG(2) << __func__ << " : "
-            << "Loaded cached skus credentials";
-    SetPurchasedState(PurchasedState::PURCHASED);
+  if (!has_credential) {
+    // TODO: we can show logic for person to login
+    // NOTE: we might save (to profile) if person EVER had a valid
+    // credential. If so, we may want to show an expired dialog
+    // instead of the "purchase" dialog.
+    skus_credential_ = "";
+    return;
   }
+
+  // if a credential is ready, we can present it
+  skus_sdk_service_->PrepareCredentialsPresentation(
+      skus::GetDomain("vpn"), "*",
+      base::BindOnce(&BraveVpnServiceDesktop::OnPrepareCredentialsPresentation,
+                     base::Unretained(this)));
```

Comment:
this is not safe. `BraveVpnServiceDesktop` depends on `SkusSdkService` so `SkusSdkService` will necessarily outlive `BraveVpnServiceDesktop`. This should be fixed by converting skus_sdk_service to mojo as discussed above

Reply:
The service will always be accessed over mojo, not just for this particular case. We've been doing this with services because it prevents some of the most common code errors and potential crashes with lifetime issues and threads. WeakPtr is frequently incorrectly used because the WeakPtr doesn't get destroyed or invalidated on the correct thread. Mojo also makes things safer and simpler to access from java and soon from rust.

Reply:
Java, Rust, cross-process - all of that is perfectly fine. I'm just a little concerned about using using mojo in the same process just to solve lifetime/threading issues. We shouldn't encourage this approach for all services, but for some I guess it's okay.

Reply:
+1 to @goodov, I'don't quite follow why should we use mojo just to replace weak pointers.
Fine if this for android compatibility or something like that

Reply:
@iefremov we discussed this in slack and we all agreed. WeakPtr is not sufficient in itself because it's easy to misuse (and was misused extensively in the past) in ways that make it unsafe when multiple threads are involved. Mojo just drops callbacks if the other end disconnects and it automatically dispatches to the correct thread. The only exception we discussed was things that are very latency sensitive like adblock. I think we should encourage this approach for most services because it is inherently safer.

Reply:
and really, how often are we not integrating with android, ios, webui and/or rust? We're not going to replace WeakPtr with mojo everywhere, the idea is to make the public interface to the service safer and simpler to integrate with. Wallet uses raw pointers internally, but that code is (should) be written by people with a good understanding of the internals.

Reply:
I don't remember I personally agreed to use mojo with keyed services just for managing lifetimes. The fact that `WeakPtr` is tricky to use with multiple threads is not a reason for that, b/c we don't actually write multithreaded services that often (I'd say, almost never).

>  I think we should encourage this approach for most services because it is inherently safer.
I disagree, b/c this is simply overkill. Most services are not multi-sequenced and not exposed to WebUIs or anywhere. Do you mean that anytime you see `WeakPtr` in a keyed service the plan is to advice rewriting it using a mojo interface?  

I agree that wallet is a good case for the mojo approach, but not normal simple services like GeminiService, DebounceService, FederatedLearning, etc

Reply:
@iefremov as I said it's not strictly for the purpose of managing lifetimes, it's also for integrating with java/rust/swift/webui and does make some otherwise tricky communication between threads much simpler when we do have that. I'm not necessarily suggesting that every single service should use it, but a lot of what we do would benefit from it.

Reply:
and in this particular case we are already exposing it for other reasons, I'm not asking him to make the service mojo just for this one case. It's already using mojo so we should use mojo here too

Reply:
Fixed in https://github.com/brave/brave-core/pull/10847/commits/f1fd81200a1006b5c0971b6805968acf4934c79a - VPN service is now getting a pending remote and then binding it. Since it owns the remote, I believe this will be OK

Reply:
ok, fine!